### PR TITLE
DOC: explain non-linear scales and imshow (closes #7661)

### LINKED ISSUE #7661: document that imshow now respects scale
With the introduction of matplotlib 2.0.0 using `plt.imshow()` together with `plt.yscale('log')` changed its behavior.

The following example code:
```python
import numpy as np
import matplotlib.pyplot as plt

tmp = np.random.random((5,5))
plt.imshow(tmp, aspect="auto", extent=(0,5, 1e0, 1e5), interpolation='nearest')
plt.yscale("log")

plt.show()
```
produces with python 3.4.3 and **matplotlib 1.5.1**
plt_imshow_2

*and the following warning:*
```shell
../.local/lib/python3.4/site-packages/matplotlib/image.py:375: UserWarning: Images are not supported on non-linear axes.
  warnings.warn("Images are not supported on non-linear axes.")
```

With python 3.4.3 and **matplotlib 2.0.0rc1** it produces
index2

Thus, 2.0.0 assumes that underneath the linear bin id of the numpy array `tmp` lies a linear data extend (y-axis) scaling.

If however my logarithmic extent is mapped correctly onto the linear bin id of the numpy array - the old behavior is exactly what represents the data. 

**Is it possible to reproduce the old behavior without relabeling the y-axis by hand?
Is there an option, the tell imshow directly that a specify data axis is already log-scaled?**
and if so:
**Is it possible to have the same behavior independent of the matplotlib version used?**

**System:**
Ubuntu 14.04.5 - Linux-Kernel 4.4.0-38
Matplotlib and Python are available via the HPC module system (probably via pip)

cc'ing: @s0vereign

### PR DESCRIPTION
**Pull request by phobson**

### CODE REVIEW

File: doc/users/whats_new/imshow_logscales.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Log-scale on image plots
+------------------------
+
+:func:`imshow` not properly displays data on log-scales.
```
Comment (afvincent) [inline doc/users/whats_new/imshow_logscales.rst]: not <- now?

Comment (phobson) [inline doc/users/whats_new/imshow_logscales.rst]: You're right on both accounts (the `if 1:` trick is there so can I can lazy when copy/pasting into jupyter -- oops)

**Comment by phobson** [comment]
Cleaned up based on comments from @afvincent and squashed/force-pushed

File: doc/users/whats_new/imshow_logscales.rst [hunk 2]
```diff
@@ -0,0 +1,31 @@
+Log-scale on image plots
```
Comment (anntzer) [inline doc/users/whats_new/imshow_logscales.rst]: "nonlinear scales"?  (also applies to logit, etc. I guess)

File: doc/users/whats_new/imshow_logscales.rst [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+------------------------
+
+:func:`imshow` now properly displays data on log-scales.
```
Comment (anntzer) [inline doc/users/whats_new/imshow_logscales.rst]: "Properly" and "correct" below don't really help describing the behavior.  Something like "draws values at the requested points in data space after application of the scale"?

**Comment by anntzer** [comment]
Actually, this should probably go into the same subheader as http://matplotlib.org/devdocs/users/whats_new.html#improved-image-support.

**Comment by phobson** [comment]
@anntzer great point

**Comment by efiring** [comment]
I'm concerned about this because I think that using imshow for anything other than linear scales is a terrible idea, confusing in the extreme.  At this stage we should strongly discourage it in the documentation, and a real improvement would be to block it in later releases after the usual warning period.

Try changing the example to:
```python
ax1.imshow(data, origin='bottom', aspect="auto", extent=(5, 0, 1e5, 1e0), interpolation='nearest')
```
and see if you think people will be happy with it, and will understand what it is doing without a long bout of head-scratching.
`imshow` simply doesn't have the right interface for anything other than normal, simple images.  Even the existing `extent` kwarg is a pain--it gives no clue as to which of the many possible sets and permutations of 4 numbers it needs.

**Comment by efiring** [comment]
Specific advice for the present PR: state that the behavior of imshow with other than linear axes has changed, but it remains undefined and should not be used.

**Comment by tacaswell** [comment]
@efiring I disagree about not using imshow on non-linear axes.  If you have evenly spaced bins it is a more efficient way than using `pcolormesh` to get the same effect.

Agree `extent` could be clearer.

**Comment by efiring** [comment]
@tacaswell, are you volunteering to explain to puzzled users why the result of the example change that I gave above looks the way it does?  And why this is an acceptable API?  We can do this right, but imshow doesn't do so, by design.  What do we lose by my suggestion of acknowledging that imshow was never designed for log scales, and now, by accident, happens to do what one might want, sometimes, with the right combination of obscure arguments?  I think this acknowledgement would be doing users a favor.

**Comment by tacaswell** [comment]
The issue with flipping extent around like that seems to be more about bottom and non-increasing values in extent being confusing as it is just as puzzilng in linear scale as with log.

```python
import numpy as np
import matplotlib.pyplot as plt

data = np.arange(30).reshape(5, 6)
x = np.linspace(0, 6, 7)
y = 10**np.linspace(0, 5, 6)
X, Y = np.meshgrid(x, y)

fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(8, 4))

ax1.imshow(data, origin='bottom', aspect="auto", extent=(6, 0, 1e5, 1e0),
           interpolation='nearest')



ax2.imshow(data, origin='bottom', aspect="auto", extent=(6, 0, 1e5, 1e0),
           interpolation='nearest')
ax2.set_yscale('log')

plt.show()

```


so

**Comment by efiring** [comment]
The log scale adds another dimension to the confusion, as illustrated by the fact that this discussion started when a user had been relying on the old behavior and considered that to be correct.

File: doc/users/whats_new/imshow_logscales.rst [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    import matplotlib.pyplot as plt
+
+    data = np.arange(30).reshape(5, 6)
+    x = np.linspace(0, 6, 7)
+    y = 10**np.linspace(0, 5, 6)
+    X, Y = np.meshgrid(x, y)
+
+    fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(8, 4))
+
+    ax1.imshow(data, aspect="auto", extent=(0, 5, 1e0, 1e5), interpolation='nearest')
```
Comment (tacaswell) [inline doc/users/whats_new/imshow_logscales.rst]: the 5 should be a 6

**Comment by tacaswell** [comment]
And the gap on the right is dependent on figure size...

**Comment by tacaswell** [comment]
In my personal use I have used `imshow` almost exclusively for data extracted from imaging detectors where the pixel value is a measure of how many photon hit that pixel (ie, a 2D histogram).  The pixel edges moving with the scale seems very natural to me.

For example

```python
import numpy as np
import matplotlib.pyplot as plt

data = np.arange(30).reshape(5, 6)
x = np.linspace(0, 6, 7)
y = 10**np.linspace(0, 5, 6)
X, Y = np.meshgrid(x, y)

fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(8, 4))
ax1.hist(np.random.rand(5000)*50000, bins=10, edgecolor='k')
ax2.hist(np.random.rand(5000)*50000, bins=10, edgecolor='k')
ax2.set_xscale('log')

plt.show()

```

so


which although it seems to expose yet another bug in log handling, seems very natural.  As does

```python
import numpy as np
import matplotlib.pyplot as plt

data = np.arange(30).reshape(5, 6)
x = np.linspace(0, 6, 7)
y = 10**np.linspace(0, 5, 6)
X, Y = np.meshgrid(x, y)

fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(8, 4))
ax1.hist(np.random.rand(5000)*100000, bins=np.logspace(0, 5, 10), edgecolor='k')
ax2.hist(np.random.rand(5000)*100000, bins=np.logspace(0, 5, 10), edgecolor='k')
ax2.set_xscale('log')

plt.show()
```
so

File: doc/users/whats_new/imshow_logscales.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    ax2.pcolormesh(x, y, np.flipud(data))
+    ax2.set_yscale('log')
+    ax2.set_title('Using ax.pcolormesh')
+    ax2.autoscale('tight')
+
+    plt.show()
+
+
+This can be understood by analogy to plotting a histogram with linearly spaced bins
+with a logarithmic x-axis.  Equal sized bins at will be displayed as wider for small
```
Comment (NelleV) [inline doc/users/whats_new/imshow_logscales.rst]: I don't understand this sentence "Equal sized bins at will be displayed …"

**Comment by tacaswell** [comment]
@efiring you mind if I do the backport of this?  I have it ready to push, but don't want mess you up if you are doing it.

**Comment by tacaswell** [comment]
backported to v2.x as cfafd3dd5

**Pull request merged**